packages/core/src/rules/oas3/no-invalid-media-type-examples.ts

```diff
@@ -50,3 +57,95 @@ export const ValidContentExamples: Oas3Rule = (opts) => {
     },
   };
 };
+
+function ensureSchemaIsResolved(
+  schema: Oas3Schema | Oas3_1Schema,
+  ctx: UserContext,
+  from?: string,
+  visited: Set<string> = new Set()
+): Oas3Schema | Oas3_1Schema | undefined {
+  if (!isRef(schema) || visited.has(schema.$ref)) {
+    return schema;
+  }
+  visited.add(schema.$ref);
+
+  const { resolve } = ctx;
+
+  const resolved = resolve<Oas3Schema | Oas3_1Schema>(schema, from);
+  if (!resolved.location) {
+    return;
+  }
+  if (!resolved.node.properties) {
+    return resolved.node;
+  }
+
+  const clonedNode = structuredClone(resolved.node);
+  if (clonedNode.properties) {
+    for (const [key, value] of Object.entries(clonedNode.properties)) {
+      if (isRef(value)) {
+        const nestedResolvedSchema = ensureSchemaIsResolved(
+          value,
+          ctx,
+          resolved.location?.source.absoluteRef,
+          visited
+        );
+        if (nestedResolvedSchema) {
+          clonedNode.properties[key] = nestedResolvedSchema;
+        } else {
+          return;
+        }
+      }
+    }
+  }
+
+  return clonedNode;
+}
+
+function isRequestExample(mediaTypeLocation: Location): boolean {
+  return (
+    mediaTypeLocation.pointer.includes('/components/requestBodies') ||
+    /paths\/[^/]+\/[^/]+\/requestBody/.test(mediaTypeLocation.pointer)
+  );
+}
+
+function isResponseExample(mediaTypeLocation: Location): boolean {
+  return (
+    mediaTypeLocation.pointer.includes('/components/responses') ||
+    /paths\/[^/]+\/[^/]+\/responses/.test(mediaTypeLocation.pointer)
```

Comment:
This is a bit fragile and hard to read. I suggest using `Responses` and `RequestBody` visitors to narrow down the context:

```ts
Responses: {
    MediaType: {
      leave(mediaType, ctx: UserContext) {
         // Filter out `writeOnly` props...
```